docs/fundamentals/connection/tls.txt

```diff
@@ -0,0 +1,202 @@
+.. _laravel-tls:
+
+========================
+Enable and Configure TLS
+========================
+
+.. facet::
+   :name: genre
+   :values: reference
+
+.. meta::
+   :keywords: code example, security, connection options, ssl
+
+.. contents:: On this page
+   :local:
+   :backlinks: none
+   :depth: 2
+   :class: singlecol
+
+Overview
+--------
+
+In this guide, you can learn how to use the TLS protocol to secure 
+your connection to a MongoDB deployment. To configure your connection to
+use TLS, enable the TLS option and provide your certificates for
+validation in your application's ``config/database.php`` file.
+
+.. tip::
+   
+   To learn more about TLS, see the Wikipedia entry on
+   :wikipedia:`Transport Layer Security <w/index.php?title=Transport_Layer_Security&oldid=1184063676>`.
+
+Enable TLS
+----------
+
+In your application's ``config/database.php`` file, you can enable TLS
+on a connection to your MongoDB deployment in one of the following ways:
+
+- Setting the ``tls`` option to ``true`` in your connection string
+- Setting the ``tls`` option to ``true`` in the ``options`` property of
+  your ``mongodb`` connection entry
+
+Select from the following :guilabel:`Connection String` and
+:guilabel:`Connection Options` tabs to see a corresponding code sample:
+
+.. tabs::
+
+   .. tab:: Connection String
+      :tabid: connection string tls true
+
+      .. code-block:: php
+         :emphasize-lines: 5
+
+         'connections' => [
+      
+             'mongodb' => [
+                 'driver' => 'mongodb',
+                 'dsn' => 'mongodb://<hostname>:<port>/?tls=true',
+                 'database' => 'myDB',
+             ]
+         ]
+
+   .. tab:: Connection Options
+      :tabid: options tls true
+      
+      .. code-block:: php
+         :emphasize-lines: 8
+
+         'connections' => [
+      
+             'mongodb' => [
+                 'driver' => 'mongodb',
+                 'dsn' => '<connection string>',
+                 'database' => 'myDB',
+                 'options'  => [
+                     'tls' => true,
+                 ],
+             ]
+         ]
+
+      To view a full list of connection options, see
+      :ref:`laravel-fundamentals-connection-options`.
+
+.. note::
+   
+   If your connection string uses a DNS SRV record by including
+   the ``mongodb+srv`` prefix, TLS is enabled on your connection by
+   default.
+
+Configure Certificates
+----------------------
+
+To successfully initiate a TLS request, your application must present 
+cryptographic certificates to prove its identity. Your application's
```

Comment:
Why is a client-side certificate necessary to initiate TLS connections to a cluster? I only expect this might be needed for [X509 authentication](https://www.mongodb.com/docs/manual/tutorial/configure-x509-client-authentication/).

Reply:
For any TLS content that I or another writer has written, we have used information from the Server docs to guide the driver documentation - https://www.mongodb.com/docs/manual/core/security-transport-encryption/#certificates

Reply:
I don't see anything in [Certificates](https://www.mongodb.com/docs/manual/core/security-transport-encryption/#certificates) that talks about clients supplying their own certificates. The introductory paragraph of that section links to [TLS/SSL Configuration for Clients](https://www.mongodb.com/docs/manual/tutorial/configure-ssl-clients/), which mainly discusses how to configure a client to validate the servers' certificates.

If this kind of content is ending up in various drivers' documentation outside of a discussion about X509, I think this needs to be revisited across the board. In any event, this does not seem relevant to PHP users as written.

Reply:
I'd still appreciate a follow-up on where this content came from, since I didn't find it on the [Certificates](https://www.mongodb.com/docs/manual/core/security-transport-encryption/#certificates) page you linked. If this is truly coming from some other docs page, I expect _that_ should also be revised.

----

I just asked about this in the drivers channel:

> Are the `tlsCertificateKeyFile` and `tlsCertificateKeyFilePassword` URI options only utilized for [X509 auth](https://www.mongodb.com/docs/manual/tutorial/configure-x509-client-authentication/#authenticate-with-a-x-509-certificate)? The [URI option spec](https://github.com/mongodb/specifications/blob/master/source/uri-options/uri-options.rst) makes no mention of X509, and the [X509 section in the auth spec](https://github.com/mongodb/specifications/blob/master/source/auth/auth.md#mongodb-x509) makes no mention of the URI options.

I'll follow up once I have a response, but in the event that `tlsCertificateKeyFile` and `tlsCertificateKeyFilePassword` are only relevant to X509 auth I'd propose removing them in favor of discussing them in some page covering auth mechanisms and not TLS connections. If anything, OCSP seems far more relevant to a docs page on TLS connections than anything pertaining to auth.

Note: the `tlsCAFile` file _is_ generally useful if an application's system is poorly configured (we've run into that before with users connecting to Atlas) and is _not_ specific to X509 auth. That can remain.

Reply:
@rcsanchez97 examined the C driver and confirmed that these URI options are only used for X509 authentication, so I reckon that supports my suggestion to remove them from general TLS docs.

Reply:
@kevinAlbs just followed up with a demonstration that these options might still be used irrespective of X509 auth: https://gist.github.com/kevinAlbs/fcdfe12af21c2cbfbd77f72aaa78a1f7.

I think the relevant server docs for this is [Set Up mongod and mongos with Client Certificate Validation](https://www.mongodb.com/docs/manual/tutorial/configure-ssl/#set-up-mongod-and-mongos-with-client-certificate-validation); however, it's not clear to me exactly what server configuration option is responsible for requiring clients to present a certificate.

I did come across the following for [`net.tls.allowConnectionsWithoutCertificates`](https://www.mongodb.com/docs/manual/reference/configuration-options/#mongodb-setting-net.tls.allowConnectionsWithoutCertificates):

> By default, the server bypasses client certificate validation unless the server is configured to use a CA file. If a CA file is provided, the following rules apply:
>
> * For clients that don't provide certificates, [mongod](https://www.mongodb.com/docs/manual/reference/program/mongod/#mongodb-binary-bin.mongod) or [mongos](https://www.mongodb.com/docs/manual/reference/program/mongos/#mongodb-binary-bin.mongos) encrypts the TLS/SSL connection, assuming the connection is successfully made.
> * For clients that present a certificate, [mongos](https://www.mongodb.com/docs/manual/reference/program/mongos/#mongodb-binary-bin.mongos) or [mongod](https://www.mongodb.com/docs/manual/reference/program/mongod/#mongodb-binary-bin.mongod) performs certificate validation using the root certificate chain specified by [CAFile](https://www.mongodb.com/docs/manual/reference/configuration-options/#mongodb-setting-net.tls.CAFile) and reject clients with invalid certificates.

That suggests [`net.tls.CAFile`](https://www.mongodb.com/docs/manual/reference/configuration-options/#mongodb-setting-net.tls.CAFile) is responsible, although the first bullet is unclear as it suggests a connection might still be made.

This likely needs clarification from someone on the server team.

But just coming back to this PR, even if we leave the options documented in this section I take issue with the following lines that suggest clients must always present certificates to connect via TLS:

> To successfully initiate a TLS request, your application must present 
cryptographic certificates to prove its identity.

and

> The following list describes the components that your client must
present to establish a TLS-enabled connection:

Reply:
While we wait for clarification from the Server team, would it be sufficient to change the phraseage in those sentences/potentially other places to be less affirmative? Something like:

To successfully initiate a TLS request, your application might need to present
cryptographic certificates to prove its identity.

Reply:
Replacing "must" works for me.